js/src/signed-varint.js

```diff
@@ -0,0 +1,127 @@
+// @flow
+
+// Copyright 2016 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+import * as Bytes from './bytes.js';
+
+export const maxVarintLength = 10;
+
+const mathPowTwoThirtyTwo = Math.pow(2, 32);
+
+function toUint32(n: number): number {
+  return n >>> 0;
+}
+
+/**
+ * Encodes `val` as signed varint and writes that into `buf` at `offset`. This returns the number
+ * of bytes written.
+ */
+export function encode(val: number, buf: Uint8Array, offset: number): number {
+  const val2 = val >= 0 ? val : -val;
+  let hi = toUint32(val2 / mathPowTwoThirtyTwo);
+  let lo = toUint32(val2);
+  // Shift left 1
+  // Get the highest n bits of lo
+  const carry = lo >>> (32 - 1);
+  lo = toUint32(lo << 1);
+  hi = (hi << 1) | carry;  // no way that this can turn negative.
+  if (val < 0) {
+    if (lo !== 0) {
+      lo--;
+    } else {
+      hi--;
+      lo = 0xffffffff;
+    }
+  }
+
+  function append(num, start, end, arr, j) {
+    for (let i = start; i < end; i += 7) {
+      if (j !== 0) {
+        arr[j - 1] |= 0x80;
+      }
+      arr[j++] = (num & (0x7f << i)) >>> i;
```

Comment:
I think there is still a lot of room for improvement here but at least it works now.